### PR TITLE
Fix godep-save by ignoring cmd/cluster-capacity and cmd/service-catalog dirs.

### PR DESCRIPTION
**Pull request by aveshagarwal**
@derekwaynecarr @eparis @stevekuznetsov @jpeeler 

Fix https://github.com/openshift/origin/issues/14751

### CODE REVIEW

File: hack/godep-save.sh [hunk 1]
```diff
@@ -56,7 +56,13 @@ REQUIRED_BINS=(
 )
 
 TMPGOPATH=`mktemp -d`
-trap "rm -rf $TMPGOPATH" EXIT
+TMPCMDREPOS=`mktemp -d`
```
Comment (stevekuznetsov) [inline hack/godep-save.sh]: Don't use deprecated back-tick expansions, instead:

    TMPGOPATH="$( mktemp -d )"

Comment (stevekuznetsov) [inline hack/godep-save.sh]: We put temporary dirs under `${BASETMPDIR}`:

    TMPGOPATH="$( mktemp --directory --tmpdir="${BASETMPDIR}" )"

File: hack/godep-save.sh [hunk 2]
```diff
@@ -56,7 +56,13 @@ REQUIRED_BINS=(
...
+function finish {
+    mv $TMPCMDREPOS/* cmd/
```
Comment (stevekuznetsov) [inline hack/godep-save.sh]: Always use braces and quote expansions:

```sh
mv "${TMPCMDREPOS}"/* cmd/
```

Comment (stevekuznetsov) [inline hack/godep-save.sh]: and likewise for other additions

**Comment by eparis** [comment]
uggh.  Could we do it by making REQUIRED_BINS not have ./... and instead have:
```
[things and stuffs]
pkg/...
api/...
assets/...
contrib/...
 docs/...
examples/...
hack/...
images/...
pkg/...
test/...
tools/...
cmd/dockerregistry/...
cmd/gitserver/...
cmd/kubefed/...
cmd/oc/...
cmd/openshift/...
```

I'm sure there is some way to generate that. But just using that list will tell us if it works...

**Comment by aveshagarwal** [comment]
>Could we do it by making REQUIRED_BINS not have ./... and instead have:

Though I understand its not everyday that we are adding top level dirs that should be handled by godep-save, but whenever it is done, that means remembering updating hack/godep-save.sh.

**Comment by aveshagarwal** [comment]
> Could we do it by making REQUIRED_BINS not have ./... and instead have:

Also do we need to include files too?

**Comment by aveshagarwal** [comment]
@eparis @stevekuznetsov updated PTAL.

File: hack/godep-save.sh [hunk 3]
```diff
@@ -164,8 +163,13 @@ fork-with-fake-packages github.com/docker/docker \
 # kubernetes tests, we have to extract the missing test dependencies and run godep-save again
 # until we converge. Because we rsync kubernetes itself above, two iterations should be enough.
 MISSING_TEST_DEPS=""
+ALL_DIRS="$(os::util::list_go_src_files | cut -d '/' -f 1-2 | sort -u | grep -v "^./cmd"|grep -v doc.go)"
+ALL_DIRS+=" $(os::util::list_go_src_files | grep "^./cmd/"| cut -d '/' -f 1-3 | LC_ALL=C sort -u|grep -v doc.go)"
```
Comment (stevekuznetsov) [inline hack/godep-save.sh]: We're doing this three times now -- let's make a function for it, slap it into `hack/lib/util/misc.sh` and not dupe it

Comment (aveshagarwal) [inline hack/godep-save.sh]: This one is a bit different than the one in verify-govet.go as it also ignores doc.go (`grep -v doc.go`).

File: hack/godep-save.sh [hunk 4]
```diff
@@ -164,8 +163,13 @@ fork-with-fake-packages github.com/docker/docker \
 # kubernetes tests, we have to extract the missing test dependencies and run godep-save again
 # until we converge. Because we rsync kubernetes itself above, two iterations should be enough.
 MISSING_TEST_DEPS=""
+ALL_DIRS="$(os::util::list_go_src_files | cut -d '/' -f 1-2 | sort -u | grep -v "^./cmd"|grep -v doc.go)"
```
Comment (eparis) [inline hack/godep-save.sh]: For predictability please use `LC_ALL=C` (agree it shouldn't matter)

Comment (eparis) [inline hack/godep-save.sh]: The `doc.go` scares me a little. Would this work? What if other .go files show up in the root dir?
```
ALL_DIRS="$(os::util::list_go_src_files | cut -d '/' -f 1-2 | grep -v ".go$" | grep -v "^./cmd" | LC_ALL=C sort -u)
```

File: hack/godep-save.sh (hunk 3 above)
Comment (eparis) [inline hack/godep-save.sh]: doc.go worries me similarly. But, that also doesn't seem to do what we need. Since it isn't excluding cc and sc.

Comment (stevekuznetsov) [inline hack/godep-save.sh]: OK, so cut out the part that is shared maybe? There's too many pipes in general and too much shared for my liking -- suuuuper easy to have a visually indistinguishable change that breaks something between the three this way

Comment (aveshagarwal) [inline hack/godep-save.sh]: @eparis what do you mean it is not excluding cc and sc?  `os::util::list_go_src_files` already excludes cc and sc.

File: hack/godep-save.sh (hunk 4 above)
Comment (aveshagarwal) [inline hack/godep-save.sh]: If you are concerned about other .go files show up in the root dir in future, why do want to have `grep -v ".go$"`? Thats why I was only including doc.go specifically not *.go.

Comment (aveshagarwal) [inline hack/godep-save.sh]: Another thing to consider what if `buildable` *.go show up?

File: hack/godep-save.sh (hunk 3 above)
Comment (eparis) [inline hack/godep-save.sh]: ohhhhhh, I'm an idiot.

File: hack/godep-save.sh (hunk 4 above)
Comment (eparis) [inline hack/godep-save.sh]: Then you'd end up with `buildable.go/...` in REQUIRED_ALL_DIRS, which would be bad.

**Comment by aveshagarwal** [comment]
@eparis @stevekuznetsov  updated PTAL.

@stevekuznetsov I am not sure how to really do what you are suggesting as only 2 places I see these changes: `verify-govet.sh` and `godep-save.sh`, as not sure to find a common stuff that really helps. Do you have any suggestion?

**Comment by stevekuznetsov** [comment]
Not sure vetting and linting needs to care about the top-level `doc.go` either so it seems like both places could use the same manipulations to `os::util::list_go_src_files`

**Comment by aveshagarwal** [comment]
>Not sure vetting and linting needs to care about the top-level doc.go either so it seems like both places could use the same manipulations to os::util::list_go_src_files

Its just not doc.go now (though those are the ones that exists today) but any *.go in top level origin/ and cmd/ dirs. Are we ok now to remove them from vetting and linting?

**Comment by stevekuznetsov** [comment]
> 

```

**Comment by stevekuznetsov** [comment]
> 

```

**Comment by stevekuznetsov** [comment]
I am very confident that there will never be any file other than the single `doc.go` at the top level or directly under `cmd/`.

**Comment by aveshagarwal** [comment]
@stevekuznetsov @eparis updated PTAL.

File: hack/lib/util/misc.sh [hunk 5]
```diff
@@ -233,3 +233,20 @@ function os::util::list_go_src_files() {
...
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+function os::util::list_go_src_dirs() {
+        local ALL_DIRS="$(os::util::list_go_src_files | cut -d '/' -f 1-2 | grep -v ".go$" | grep -v "^./cmd" | LC_ALL=C sort -u)"
+        ALL_DIRS+=" $(os::util::list_go_src_files | grep "^./cmd/"| cut -d '/' -f 1-3 | grep -v ".go$" | LC_ALL=C sort -u)"
+        echo "${ALL_DIRS}"
```
Comment (stevekuznetsov) [inline hack/lib/util/misc.sh]: No need for `echo` or for collecting in a variable:

```sh
function os::util::list_go_src_dirs() {
    os::util::list_go_src_files | cut -d '/' -f 1-2 | grep -v ".go$" | grep -v "^./cmd" | LC_ALL=C sort -u
    os::util::list_go_src_files | grep "^./cmd/"| cut -d '/' -f 1-3 | grep -v ".go$" | LC_ALL=C sort -u
}
```

**Comment by aveshagarwal** [comment]
@stevekuznetsov updated PTAL.

**Comment by eparis** [comment]
seems good to me...
@aveshagarwal will you remind us all to merge on monday? (or is this blocking you on something you must get in today/tomorrow)

**Comment by aveshagarwal** [comment]
@eparis sure.

**Comment by aveshagarwal** [comment]
@eparis @stevekuznetsov Reminder for merge tag.

**Comment by stevekuznetsov** [comment]
[merge]

**Pull request merged**